### PR TITLE
fix saving and restoring checkpoint from dlog

### PR DESCRIPTION
**Pull request by simingweng**
flush the output stream before closing the underlying dlog writer
return -1 when no byte is read from the dlog reader

### CODE REVIEW

**Comment by simingweng** [comment]
@nwangtw please add reviewer.

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java [hunk 1]
```diff
@@ -120,7 +99,7 @@ public int read(byte[] b, int off, int len) throws IOException {
     if (currentLogRecord == null) {
       currentLogRecord = nextLogRecord();
       if (currentLogRecord == null) {
-        return read;
+        return -1;
```
Comment (nwangtw) [inline heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: The original logic seems to return a 0 and then return -1 in the next run if it is EOS. It seems to be safer to me (if currently there is no data but it is not EOS yet). Is there any issue?

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java [hunk 2]
```diff
@@ -6,9 +6,9 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
- *
- *   http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p>
```
Comment (nwangtw) [inline heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: Please revert this change, to be consistent to all other JAVA files.

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java (hunk 1 above)
Comment (simingweng) [reply heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: Yes, the protocol buffer message parser insists on not seeing any 0 return value, therefore the first 0 already results in an error. We have to return -1 immediately when the end is reached. I tested this on my Heron+Bookkeeper on k8s cluster with a stateful topology.

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java (hunk 2 above)
Comment (simingweng) [reply heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: Ok, I saw the Travis CI build failed because of this, will revert accordingly.

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java (hunk 1 above)
Comment (nwangtw) [reply heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: kk. thx

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java (hunk 2 above)
Comment (nwangtw) [reply heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: thx

File: heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java [hunk 3]
```diff
@@ -100,6 +60,25 @@ private static LogRecordWithInputStream nextLogRecord(LogReader reader) throws I
...
+   * Get input stream representing next entry in the
+   * ledger.
+   *
+   * @return input stream, or null if no more entries
+   */
+  private LogRecordWithInputStream nextLogRecord() throws IOException {
+    try {
+      return nextLogRecord(reader);
+    } catch (EndOfStreamException e) {
+      eos = true;
```
Comment (nwangtw) [inline heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: Could it be helpful to have a log here?

Comment (simingweng) [reply heron/io/dlog/src/java/org/apache/heron/dlog/DLInputStream.java]: I didn’t really change this part of the file, it was my IDE automatically re-orgonized the code. But, I do think log here is helpful. Will update the pull request.

**Comment by nwangtw** [comment]
It seems like unit tests are failing:

50/50 files checked 100% done
(21:28:03) INFO: Elapsed time: 1264.383s, Critical Path: 154.04s
(21:28:03) INFO: 1192 processes: 1192 local.
(21:28:03) INFO: Build completed, 2 tests FAILED, 1547 total actions
//heron/io/dlog/tests/java:DLInputStreamTest                             FAILED in 1.2s
    FAILED  org.apache.heron.dlog.DLInputStreamTest.testReadEos (0.0s)
//heron/statefulstorages/tests/java:DlogStorageTest                      FAILED in 5.2s
    FAILED  org.apache.heron.statefulstorage.dlog.DlogStorageTest.testRestore (0.1s)
Test cases: finished with 1094 passing and 2 failing out of 1096 test cases

**Comment by simingweng** [comment]
> It seems like unit tests are failing:
> 
> 50/50 files checked 100% done
> (21:28:03) INFO: Elapsed time: 1264.383s, Critical Path: 154.04s
> (21:28:03) INFO: 1192 processes: 1192 local.
> (21:28:03) INFO: Build completed, 2 tests FAILED, 1547 total actions
> //heron/io/dlog/tests/java:DLInputStreamTest FAILED in 1.2s
> FAILED org.apache.heron.dlog.DLInputStreamTest.testReadEos (0.0s)
> //heron/statefulstorages/tests/java:DlogStorageTest FAILED in 5.2s
> FAILED org.apache.heron.statefulstorage.dlog.DlogStorageTest.testRestore (0.1s)
> Test cases: finished with 1094 passing and 2 failing out of 1096 test cases

yes, I've updated the unit test to expect -1 as return value when the end of the stream is reached.

**Comment by nwangtw** [comment]
This one is still failing:

//heron/statefulstorages/tests/java:DlogStorageTest                      FAILED in 6.4s
    FAILED  org.apache.heron.statefulstorage.dlog.DlogStorageTest.testRestore (0.1s)
Test cases: finished with 1095 passing and 1 failing out of 1096 test cases

**Comment by simingweng** [comment]
> heron/statefulstorages/tests/java:DlogStorageTest

sorry, this slipped through the crack, I'm correct it.

**Comment by nwangtw** [comment]
Cool. Thx!
